### PR TITLE
#103 - Implement convex hull for an array of sets

### LINKED ISSUE #103: Implement convex hull array
Implement the symbolic convex hull array operator such that

```julia
julia> C = ConvexHullArray([Singleton([1.0, 0.5]), Singleton([1.1, 0.2]), Singleton([1.4, 0.3]), Singleton([1.7, 0.5]), Singleton([1.4, 0.8])])
```

works (alias: `CHArray`). The choice `xyzArray` follows the nomenclature used previously for `MinkowskiSum` and `CartesianProduct`.

### PR DESCRIPTION
**Pull request by mforets**
Closes #103.

### CODE REVIEW

File: src/ConvexHull.jl [hunk 1]
```diff
@@ -106,3 +107,52 @@ function σ(d::AbstractVector{<:Real}, ch::ConvexHull)::AbstractVector{<:Real}
...
+
+### Fields
+
+- `array` -- array of sets
+"""
+struct ConvexHullArray{N<:Real, S<:LazySet{N}} <: LazySet{N}
+    array::Vector{S}
+end
+# type-less convenience constructor
+ConvexHullArray(a::Vector{S}) where {S<:LazySet{N}} where {N<:Real} = ConvexHullArray{N, S}(a)
```
Comment (schillic) [inline src/ConvexHull.jl]: Is this needed when there is no inner constructor? I remember that sometimes you need it...

File: src/ConvexHull.jl [hunk 2]
```diff
@@ -106,3 +107,52 @@ function σ(d::AbstractVector{<:Real}, ch::ConvexHull)::AbstractVector{<:Real}
...
+(ConvexHullArray(∅::E, ::E)) where {E<:EmptySet} = ∅
+dim(ch::ConvexHullArray) = dim(ch.array[1])
+
+function σ(d::AbstractVector{<:Real}, ch::ConvexHullArray)::AbstractVector{<:Real}
+    s = r = similar(d)
+    for (i, ci) in enumerate(ch.array)
+       s[i] = σ(d, ci)
+       r[i] = dot(d, s[i])
+    end
+    return s[indmax(r)]
```
Comment (schillic) [inline src/ConvexHull.jl]: Are you sure about this? What about negative directions?

File: src/ConvexHull.jl (hunk 1 above)
Comment (mforets) [reply src/ConvexHull.jl]: yes, its' needed, just checked

**Comment by mforets** [comment]
i changed the algorithm for σ(d, cha) (only keep greatest value of support vector), and added some examples. the idea of the second one was the plot (but i don't want to add `using Plots`)

```julia
b = [Ball2([2*pi*i/100, sin(2*pi*i/100)], 0.01) for i in 1:100];
c = ConvexHullArray(b);
plot(c, 1e-3, alpha=0.2, color="red")
plot!(b, 1e-3, alpha=0.2, color="orange")
```

maybe i'll propose to add it to the [Convex hulls](https://juliareach.github.io/LazySets.jl/latest/man/convex_hulls.html) manual section.

**Comment by mforets** [comment]
done, this is ready to go..

**Comment by schillic** [comment]
It is a bit inconvenient that we do not have a function symbol for the convex hull. Now if you want to add a new set to a `ConvexHullArray`, you have to access the underlying array. Should we say that `CH(cha::ConvexHullArray, X::LazySet) = cha.array.add(X)` (+ the symmetric case + the *merge* case for two `ConvexHullArray`s)?

File: src/ConvexHull.jl [hunk 3]
```diff
@@ -164,6 +164,18 @@ Alias for `ConvexHullArray`.
 """
 const CHArray = ConvexHullArray
 
+CH(cha::ConvexHullArray, ∅::EmptySet) = cha
+CH(cha::EmptySet, ::ConvexHullArray) = cha
```
Comment (schillic) [inline src/ConvexHull.jl]: `cha` should be the *second* argument.

File: src/ConvexHull.jl [hunk 4]
```diff
@@ -164,6 +164,18 @@ Alias for `ConvexHullArray`.
...
+
+function CH(cha::ConvexHullArray, X::LazySet)
+    push!(cha.array.X)
```
Comment (schillic) [inline src/ConvexHull.jl]: How does this work? Is it not `push!(cha.array, X)`?

**Comment by mforets** [comment]
the `CH(cha::ConvexHullArray, X::LazySet)` requires i think to change the def of `CH` so i'm leaving this outside this issue, see #165.

**Pull request merged**